### PR TITLE
KAFKA-16335: Remove deprecated method of StreamPartitioner

### PR DESCRIPTION
**Pull request by caioguedes**
Remove deprecated method StreamPartitioner#partition.

All tests were updated if the new signature of `StreamPartitioner#partitions`.

### CODE REVIEW

**Comment by mjsax** [comment]
Next release is 3.8, so we cannot merge this yet. We can of course leave this PR open for the time being, and I can review it when we start 4.0 release (which is planned after 3.8).

**Comment by caioguedes** [comment]
Thanks @mjsax, I will fix the problems in the CI and let it open. 😄

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java [hunk 1]
```diff
@@ -1071,14 +1071,14 @@ public <VR, KO, VO> KTable<K, VR> leftJoin(final KTable<KO, VO> other,
         return doJoinOnForeignKey(other, foreignKeyExtractor, joiner, TableJoined.with(null, null), materialized, true);
     }
 
-    private final Function<Optional<Set<Integer>>, Integer> getPartition = maybeMulticastPartitions -> {
-        if (!maybeMulticastPartitions.isPresent()) {
-            return null;
+    private final Function<Optional<Set<Integer>>, Optional<Set<Integer>>> getPartition = maybeMulticastPartitions -> {
+        if (maybeMulticastPartitions == null || !maybeMulticastPartitions.isPresent()) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java]: Do we need this `null` check? Seems it's actually not allowed to return `null`?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java [hunk 2]
```diff
@@ -1229,7 +1229,7 @@ private <VR, KO, VO> KTable<K, VR> doJoinOnForeignKey(final KTable<KO, VO> forei
 
         final StreamPartitioner<K, SubscriptionResponseWrapper<VO>> foreignResponseSinkPartitioner =
                 tableJoinedInternal.partitioner() == null
-                        ? (topic, key, subscriptionResponseWrapper, numPartitions) -> subscriptionResponseWrapper.getPrimaryPartition()
+                        ? (topic, key, subscriptionResponseWrapper, numPartitions) -> Optional.of(Collections.singleton(subscriptionResponseWrapper.getPrimaryPartition()))
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java]: Should we change `getPrimaryPartition` to return an `Optional` instead ?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java [hunk 3]
```diff
@@ -153,7 +153,7 @@ public <K, V> void send(final String topic,
             }
             if (partitions.size() > 0) {
                 final Optional<Set<Integer>> maybeMulticastPartitions = partitioner.partitions(topic, key, value, partitions.size());
-                if (!maybeMulticastPartitions.isPresent()) {
+                if (maybeMulticastPartitions == null || !maybeMulticastPartitions.isPresent()) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java]: As above.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java [hunk 4]
```diff
@@ -1569,7 +1569,7 @@ private void assertNextOutputRecord(final TestRecord<String, String> record,
     }
 
     private StreamPartitioner<Object, Object> constantPartitioner(final Integer partition) {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java]: Can we improve the generic types?

**Review by mjsax — COMMENTED** [review]
Thanks for the PR. `trunk` was bumped to `4.0-SNAPSHOT` so we can finally work on this.

Make a pass. -- Could you also rebase this PR to `trunk` to make sure it's still up-to-date? We cannot rely on the old Jenkins run any longer.

**Comment by caioguedes** [comment]
> Thanks for the PR. `trunk` was bumped to `4.0-SNAPSHOT` so we can finally work on this.
> 
> Make a pass. -- Could you also rebase this PR to `trunk` to make sure it's still up-to-date? We cannot rely on the old Jenkins run any longer.

Thanks @mjsax for reviewing the PR, I'll try to work in the changes this days, so I might let you know if any question comes up.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java (hunk 2 above)
Comment (caioguedes) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java]: Sounds good, should we also create an additional constructor dropping the parameter `primaryPartition`? I see many calls of the actual constructor passing `null`? Something like this?

```java
// New Additional Constructor
public SubscriptionResponseWrapper(final long[] originalValueHash, final FV foreignValue) {
        this(originalValueHash, foreignValue, CURRENT_VERSION, null);
}

// Other Constructors
public SubscriptionResponseWrapper(final long[] originalValueHash, final FV foreignValue, final Integer primaryPartition) {
        this(originalValueHash, foreignValue, CURRENT_VERSION, primaryPartition);
}

public SubscriptionResponseWrapper(
        final long[] originalValueHash,
        final FV foreignValue,
        final byte version,
        final Integer primaryPartition) {
        // ... code omitted 
        this.primaryPartition = Optional.ofNullable(primaryPartition);
}
```

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java (hunk 4 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java]: @caioguedes Any thoughts?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java (hunk 2 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java]: I don't think it's worth to add an overloaded constructor. This adds more noise than it simplifies code IMHO.

However, looking into the code mode closely, it seems that `getPrimaryPartition ` is called elsewhere, so changing the return type to `Optional` might actually not be ideal?

However, we need to take into account that `getPrimaryPartition` could return `null` for which we should not return a singleton collection with `null` in it here, but rather an `Optional.empty()` (please correct me if I am wrong).

**Review by mjsax — COMMENTED** [review]
Sorry for late reply... Let's push this over the finish line.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java (hunk 4 above)
Comment (caioguedes) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java]: Sure, most of the test cases are using `StringDeserializer`, it makes sense move to `StreamPartitioner<String, String>`

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java (hunk 2 above)
Comment (caioguedes) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java]: @mjsax, about the overloaded, let put it aside, as you mentioned better keep it simple.

About `getPrimaryPartition` we should use `Optional.empty()` in this case, and `Optional.ofNullable()` for the constructor parameter.

However about the returning type change in `getPrimaryPartition`, I've noticed that `SubscriptionResponseWrapper` is serialized, do we should concern about the type change in this case?

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java]: Yes, let's keep the return type of `getPrimaryPartition` unchanged.

However, if `getPrimaryPartition` returns `null`, we should not construct `Optional.of(Collections.singleton(null)` here, but rather `Optional.empty()`, right? So we need to change the code here accordingly?

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java]: As a matter of fact, we have a failing test that I believe exactly exposes this bug:
```
java.lang.NullPointerException
	at org.apache.kafka.streams.processor.internals.RecordCollectorImpl.send(RecordCollectorImpl.java:163)
	at org.apache.kafka.streams.processor.internals.RecordCollectorTest.shouldUseDefaultPartitionerAsPartitionReturnsNull(RecordCollectorTest.java:606)
```

Comment (caioguedes) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java]: Just pushed the change :)

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java [hunk 5]
```diff
@@ -570,7 +570,7 @@ public Optional<Set<Integer>> partitions(final String topic, final String key, f
     public void shouldUseDefaultPartitionerAsPartitionReturnsNull() {
 
         final StreamPartitioner<String, Object> streamPartitioner =
-                (topic, key, value, numPartitions) -> null;
+                (topic, key, value, numPartitions) -> Optional.empty();
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java:575]: The test name should be updated... we used `null` for the old method, but `empty` for the new one, which we test now.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java [hunk 6]
```diff
@@ -1230,7 +1230,7 @@ private <VR, KO, VO> KTable<K, VR> doJoinOnForeignKey(final KTable<KO, VO> forei
 
         final StreamPartitioner<K, SubscriptionResponseWrapper<VO>> foreignResponseSinkPartitioner =
                 tableJoinedInternal.partitioner() == null
-                        ? (topic, key, subscriptionResponseWrapper, numPartitions) -> Optional.of(Collections.singleton(subscriptionResponseWrapper.getPrimaryPartition()))
+                        ? (topic, key, val, numPartitions) -> val.getPrimaryPartition() == null ? Optional.empty() : Optional.of(Collections.singleton(val.getPrimaryPartition()))
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java]: Why did you rename `subscriptionResponseWrapper` to `val`? The old name seem to be much more descriptive.

Nit: should we call `val.getPrimaryPartition()` twice as the code does now, or better introduce a variable which stores the result and only call it once?

Comment (caioguedes) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java]: Yeah, you are right. I'm keeping the name and also extract to a property.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java (hunk 5 above)
Comment (caioguedes) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java:575]: Done :)

**Comment by mjsax** [comment]
Thanks for the PR @caioguedes -- merged to `trunk`.
